packages/firestore/src/core/sync_engine.ts

```diff
@@ -979,14 +616,457 @@ async function synchronizeViewAndComputeSnapshot(
+    const event = new RemoteEvent(
+      SnapshotVersion.min(),
+      /* targetChanges= */ new Map<TargetId, TargetChange>(),
+      /* targetMismatches= */ new SortedSet<TargetId>(primitiveComparator),
+      documentUpdates,
+      resolvedLimboDocuments
+    );
+
+    await applyRemoteEvent(syncEngineImpl, event);
+
+    // Since this query failed, we won't want to manually unlisten to it.
+    // We only remove it from bookkeeping after we successfully applied the
+    // RemoteEvent. If `applyRemoteEvent()` throws, we want to re-listen to
+    // this query when the RemoteStore restarts the Watch stream, which should
+    // re-trigger the target failure.
+    syncEngineImpl.activeLimboTargetsByKey = syncEngineImpl.activeLimboTargetsByKey.remove(
+      limboKey
+    );
+    syncEngineImpl.activeLimboResolutionsByTarget.delete(targetId);
+    pumpEnqueuedLimboResolutions(syncEngineImpl);
+  } else {
+    await syncEngineImpl.localStore
+      .releaseTarget(targetId, /* keepPersistedTargetData */ false)
+      .then(() => removeAndCleanupTarget(syncEngineImpl, targetId, err))
+      .catch(ignoreIfPrimaryLeaseLoss);
+  }
+}
+/**
+ * Registers a view for a previously unknown query and computes its initial
+ * snapshot.
+ */
+async function initializeViewAndComputeSnapshot(
+  syncEngineImpl: SyncEngineImpl,
+  query: Query,
+  targetId: TargetId,
+  current: boolean
+): Promise<ViewSnapshot> {
+  // PORTING NOTE: On Web only, we inject the code that registers new Limbo
+  // targets based on view changes. This allows us to only depend on Limbo
+  // changes when user code includes queries.
+  syncEngineImpl.docsChangeListener = (queryView, changes, remoteEvent) =>
+    applyDocChanges(syncEngineImpl, queryView, changes, remoteEvent);
+
+  const queryResult = await syncEngineImpl.localStore.executeQuery(
+    query,
+    /* usePreviousResults= */ true
+  );
+  const view = new View(query, queryResult.remoteKeys);
+  const viewDocChanges = view.computeDocChanges(queryResult.documents);
+  const synthesizedTargetChange = TargetChange.createSynthesizedTargetChangeForCurrentChange(
+    targetId,
+    current && syncEngineImpl.onlineState !== OnlineState.Offline
+  );
+  const viewChange = view.applyChanges(
+    viewDocChanges,
+    /* updateLimboDocuments= */ syncEngineImpl.isPrimaryClient,
+    synthesizedTargetChange
+  );
+  updateTrackedLimbos(syncEngineImpl, targetId, viewChange.limboChanges);
+
+  debugAssert(
+    !!viewChange.snapshot,
+    'applyChanges for new view should always return a snapshot'
+  );
+
+  const data = new QueryView(query, targetId, view);
+
+  syncEngineImpl.queryViewsByQuery.set(query, data);
+  if (syncEngineImpl.queriesByTarget.has(targetId)) {
+    syncEngineImpl.queriesByTarget.get(targetId)!.push(query);
+  } else {
+    syncEngineImpl.queriesByTarget.set(targetId, [query]);
+  }
+
+  return viewChange.snapshot;
+}
+
+async function applyDocChanges(
```

Comment:
`emitNewSnapsAndNotifyLocalStore` is not a free standing function (yet). It likely won't need to be one for size savings (since most codepaths rely on it), but we can make it one for consistency.

Reply:
We might as well just adopt the one style so you don't have to think about whether a given operation is a method or a free function.